File: scripts/parse.py
#!/usr/bin/env python3

from bs4 import BeautifulSoup
import json
import re
import os
import time

list=[]

def find_h(file):
    for n in range(5):
        for h in soup.find_all('h'+str(n)):
            if h.has_attr('id'):
                id = h.attrs['id']
                list.append({
                    'text': h.get_text(),
                    'tag': h.name,
                    'id': id,
                    'href': '#' + id,
                    'file': file
                })

def find_span(file):
    for span in soup.find_all('span'):
        if span.has_attr('id'):
            id = span.attrs['id']
            list.append({
                'text': span.get_text(),
                'tag': span.name,
                'id': id,
                'href': '#' + id,
                'file': file
            })

def find_a(file):
    for a in soup.find_all('a'):
        if a.has_attr('href'):
            list.append({
                'text': a.get_text(),
                'tag': a.name,
                'href': a.attrs['href'],
                'file': file
            })


def TimeStampToTime(timestamp, format):
    timeStruct = time.localtime(timestamp)
    return time.strftime(format, timeStruct)


def get_FileCreateTime(file, format='%Y-%m-%d %H:%M:%S'):
    # filePath = unicode(filePath, 'utf8')
    t = os.path.getctime(file)
    return TimeStampToTime(t, format)


timestamps={}

for file in os.listdir("./posts/"):
    if file.endswith('html'):
        soup = BeautifulSoup(open("./posts/" + file), 'html.parser')
        date = soup.find_all('')
        timestamps[file] = {
            "timestamp": get_FileCreateTime("./posts/" + file),
            "month": get_FileCreateTime("./posts/" + file, "%Y-%m"),
            "title": soup.head.title.get_text(),
            "file": file
        }
        find_h(file)
        find_span(file)
        find_a(file)


for index in range( len( list ) ):
    item = list[index]
    text = item['text']
    text = re.sub(r'\n', '', text)
    text = re.sub(r'\s+', ' ', text)
    item['text'] = text.strip() or item['href']
    if item['text'] == '' and item['href'] != '':
        item['text'] = item['href']
    if len(item['text']) > 0:
        # item['text'] = '[' + item['file'] + ']' + item['text']
        item['value'] = item['text']
        if item['href'].startswith('#'):
            item['href'] = item['file'] + item['href']
        item['link'] = item['href']
        list[index] = item

try:
   f = open("./assets/js/dist/stats.js", 'w', encoding = 'utf-8')
   f.write("window.$stats="+json.dumps(list)+";window.$timestamp="+json.dumps(timestamps))
   # perform file operations
finally:
   f.close()

File: scripts/parse.js
/** jsx?|tsx? file header */

const fs = require('fs')
const HTMLParser = require('node-html-parser')

const add0 = (s) => (String(s).length === 1 ? '0' + s : s)
function formatTime(str, format = 'YYYY-MM-DD hh:mm:ss') {
  const d = new Date(str)
  const year = d.getFullYear()
  const m = d.getMonth() + 1
  const day = d.getDate()
  const h = d.getHours()
  const ms = d.getMinutes()
  const ss = d.getSeconds()

  return format
    .replace(/yyyy/i, year)
    .replace('MM', add0(m))
    .replace(/dd/i, add0(day))
    .replace('hh', add0(h))
    .replace('mm', add0(ms))
    .replace('ss', add0(ss))
}

function parseHTMLFiles() {
  return new Promise((resolve, reject) => {
    fs.readdir('./posts/', (err, files) => {
      if (err) {
        reject(err)
        return
      }

      const ps = []
      files.map((file) => {
        if (/^\./.test(file)) {
          return
        }
        ps.push(
          new Promise((resolve, reject) => {
            const path = `./posts/${file}`
            fs.readFile(
              path,
              {
                encoding: 'utf8'
              },
              (err, data) => {
                if (err) {
                  reject(err)
                  return
                }
                const html = data.toString()
                const root = HTMLParser.parse(html)
                fs.stat(path, (err, stats) => {
                  if (!err) {
                    const { ctime, birthtime } = stats
                    resolve({
                      file,
                      root,
                      stats: {
                        ctime,
                        birthtime,
                        updatedAt: formatTime(ctime),
                        createdAt: formatTime(birthtime)
                      }
                    })
                  }
                })
              }
            )
          })
        )
      })
      Promise.all(ps).then(resolve, reject)
    })
  })
}

function trim(s) {
  return s
    .replace(/\n/g, '')
    .replace(/\s+/g, ' ')
    .replace(/^\d[.\d\s]+/g, '')
}

function findIDLinks(root, filename) {
  return root
    .querySelectorAll('span[id],h2[id],h3[id],h4[id]')
    .map((e) => {
      const o = {
        text: trim(e.text),
        tagName: e.tagName,
        ...e.attributes,
        filename
      }

      if (!o.id || !o.text) return null

      const tag = e.querySelector('span.tag')
      if (tag) {
        o.tags = tag
          .querySelectorAll('span')
          .map((t) => t && t.text)
          .filter(Boolean)
      }
      return o
    })
    .filter(Boolean)
}

function findALinks(root, filename) {
  return root
    .querySelectorAll('a')
    .map((a) => {
      const item = {
        text: trim(a.text),
        url: a.attributes.href,
        tag: a.tagName,
        filename
      }

      if (item.text === ' ') {
        item.text = item.url
      }

      return item
    })
    .filter((a) => a.url && a.text)
}

function findTOC(root, file) {
  if (file !== 'home.html') {
    return null
  }
  return root
    .querySelector('#table-of-contents')
    .querySelectorAll('ul>li')
    .map((li) => {
      const a = li.querySelector('a')
      const { href } = a.attributes
      return {
        text: trim(a.text),
        url: href
      }
    })
    .filter((a) => a.url && a.text)
}

function parseRoots(roots) {
  const stats = {}
  roots.forEach((item) => {
    const name = item.file
    if (stats[name] == null) {
      stats[name] = {}
    }
    const file = stats[name]
    file.IDLinks = findIDLinks(item.root, name)
    file.aLinks = findALinks(item.root, name)
    file.toc = findTOC(item.root, name)
    fs.writeFile(
      `./assets/js/stats/${name}.js`,
      `window.$stats=window.$stats||{};window.$stats['${name}']=${JSON.stringify(
        file
      )}`,
      {
        encoding: 'utf8'
      },
      handleError
    )
  })

  fs.writeFile(
    `./assets/js/dist/stats.js`,
    `window.$pages=${JSON.stringify(
      roots.reduce((o, next) => {
        const { birthtime, ctime, updatedAt } = next.stats
        const d = new Date(birthtime)
        let month = d.getMonth() + 1,
          year = d.getFullYear(),
          day = d.getDate()

        month = add0(month)
        day = add0(day)

        let category = [],
          tags = []
        const meta = next.root.querySelector('meta[name=category]')
        if (meta) {
          category = (meta.getAttribute('content') || '').split(',')
        }
        const metaTag = next.root.querySelector('meta[name=tags]')
        if (metaTag) {
          tags = (metaTag.getAttribute('content') || '').split(',')
        }
        let createdAt = next.root.querySelector('meta[name=createdAt]')
        if (createdAt) {
          createdAt = createdAt.getAttribute('content')
        }

        console.log(next.file)
        if (checkValidPage(next)) {
          o[next.file] = {
            createdAt,
            updatedAt,
            birthtime,
            ctime,
            month,
            year,
            day,
            date: `${month}-${day}`,
            title: next.root.querySelector('.title').text,
            category,
            tags
          }
        }

        return o
      }, {})
    )}`,
    'utf8',
    handleError
  )

  return stats
}

function handleError(err) {
  if (err) {
    console.error('[Error]', err)
  }
}

parseHTMLFiles().then(parseRoots, handleError)

function checkValidPage(next) {
  if (['index.html'].includes(next.file)) return

  if (/^\./.test(next.file)) return

  // valid
  return next
}
